### ISSUE TITLE
IRCClient heartbeat

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5047]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)| @jonathanj reported|
|-|-|
|Trac ID|trac#5047|
|Type|enhancement|
|Created|2011-04-11 09:43:19Z|
|Branch|https://github.com/twisted/twisted/tree/irc-heartbeat-5047-2|

IRC servers tend to drop clients that they consider inactive, client sends no data for some period of time, causing the transport to die in a way that may not invoke the usual `connectionLost` method until it is written to.

A simple implementation could send a `PING server.hostname` periodically. 

<details><summary>Searchable metadata</summary>

```
trac-id__5047 5047
type__enhancement enhancement
reporter__jonathanj jonathanj
priority__normal normal
milestone__ 
branch__branches_irc_heartbeat_5047_2 branches/irc-heartbeat-5047-2
branch_author__jonathanj jonathanj
status__closed closed
resolution__fixed fixed
component__words words
keywords__irc irc
time__1302514999000000 1302514999000000
changetime__1307529770000000 1307529770000000
version__None None
owner__jonathanj jonathanj
cc__ralphm cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_2"></a>|@jonathanj commented|
|-|-|

(In [31623]) Branching to 'irc-heartbeat-5047'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_3"></a>|@jonathanj commented|
|-|-|

(In [31624]) Implement a heartbeat for IRCClient, refs [#5047](https://github.com/twisted/twisted/issues/5047).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_4"></a>|@jonathanj removed owner|
|-|-|

Use a `LoopingCall` to send a _PING_ message every 120 seconds, by default, to keep the IRC connection alive. The heartbeat can be disabled by setting `IRCClient.heartbeatInterval` to `None`.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @jonathanj|
|-|-|

Thanks. Minor things:

 1. New methods need an @since marker
 1. The docstring for `heartbeatInterval` doesn't mention that the default is 120 seconds.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_6"></a>|@jonathanj commented|
|-|-|

(In [31630]) Add @since markers and improve heartbeatInterval docstring, refs [#5047](https://github.com/twisted/twisted/issues/5047).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_7"></a>|@jonathanj removed owner|
|-|-|

Replying to [thijs](#note_5):
> Thanks. Minor things:
> 
>  1. New methods need an @since marker
>  1. The docstring for `heartbeatInterval` doesn't mention that the default is 120 seconds.

Done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. Yay.  I love `StringTransport`.
  1. In the `IRCClient` class docstring, probably worth mentioning that the `hostname` attribute is `None` until the server informs the client of its hostname, and then it is initialized from that server message.
  1. In `test_irc.py`:
    1. `assertIdentical` is preferred over `assertTrue(x is y)` and `assertNotIdentical` over `assertTrue(x is not y)`
    1. `clock.advance(n)` can be used instead of `clock.pump([n])`
    1. If you like, you can also assert about the length of `clock.getDelayedCalls()`.  If it's 0, then all delayed calls must have been cleaned up (this, rather than advancing the clock again and asserting that no more calls were made).
  1. A number of existing tests are failing now, some because of left over delayed calls, some because of the `connectionLost` signature change (which, ugh, what the heck is wrong with `Protocol.connectionLost`, jeez).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_10"></a>|@jonathanj set owner to @exarkun|
|-|-|

> 2. In the `IRCClient` class docstring, probably worth mentioning that the `hostname` attribute is `None` until the server informs the client of its hostname, and then it is initialized from that server message.
> 3.1. `assertIdentical` is preferred over `assertTrue(x is y)` and `assertNotIdentical` over `assertTrue(x is not y)`
> 3.2. `clock.advance(n)` can be used instead of `clock.pump([n])`
> 3.3. If you like, you can also assert about the length of `clock.getDelayedCalls()`.  If it's 0, then all delayed calls must have been cleaned up (this, rather than advancing the clock again and asserting that no more calls were made).

Done.

> 4. A number of existing tests are failing now, some because of left over delayed calls, some because of the `connectionLost` signature change (which, ugh, what the heck is wrong with `Protocol.connectionLost`, jeez).

Fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_12"></a>|@jonathanj commented|
|-|-|

(In [31990]) Branching to 'irc-heartbeat-5047-2'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_13"></a>|@jonathanj commented|
|-|-|

(In [31991]) Merge forward and resolve conflicts, refs [#5047](https://github.com/twisted/twisted/issues/5047).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. `heartbeatInterval` is really an _ivar_ not a _cvar_.  Hardly anything is a _cvar_ - basically just stuff that you actually access through the class object directly, or possibly mutable stuff that really gets mutated.
  1. `CTCPTest.setUp` makes `self.transport.loseConnection` and `self.client.connectionLost` cleanup actions, but `CTCPTest.tearDown` still calls them both.
  1. The nested `_sendHeartbeat` definition seems counterproductive.  Lose it and `self.calls` and just observe `self.transport.value()`?  That seems to be a better assertion, since it still covers the 'PING' and it verifies `IRCClient._sendHeartbeat` is what the looping call calls, instead of manually calling that method in the test method.
  1. Similarly, `_createHeartbeat` could call the original implementation and just the the `clock` attribute afterwards.

Thanks for working on this.  Please merge when you're happy with it and [buildbot is green](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/irc-heartbeat-5047-2).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_15"></a>|@jonathanj set status to `closed`|
|-|-|

(In [32066]) Merge irc-heartbeat-5047-2.

Author: jonathanj
Reviewer: exarkun, thijs
Fixes: [#5047](https://github.com/twisted/twisted/issues/5047)

Implement a heartbeat, a "PING" message to the server at regular intervals, for IRCClient to avoid having the transport die in a way that does not invoke connectionLost until written to.
